identity_core/src/common/single_struct_error.rs

```diff
@@ -0,0 +1,179 @@
+// Copyright 2020-2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::borrow::Cow;
+use std::error::Error;
+use std::fmt::Debug;
+use std::fmt::Display;
+
+/// A container implementing the [`std::error::Error`] trait.
+///
+/// Instances always carry a corresponding `kind` of type `T` and may be extended with custom error
+/// messages, source and recovery data.
+///
+/// This type is mainly designed to accommodate for the [single struct error design pattern](https://nrc.github.io/error-docs/error-design/error-type-design.html#single-struct-style).
+///
+/// When used in a specialized context it is recommended to use a type alias (i.e. `type MyError =
+/// SingleStructError<MyErrorKind>` or `type MyError = SingleStructError<MyErrorKind, MyRecoveryData>`).
+#[derive(Debug)]
+pub struct SingleStructError<T: Debug + Display, S: Debug = ()> {
```

Comment:
I took the name from this section of [the following error handling guide](https://nrc.github.io/error-docs/error-design/error-type-design.html#single-struct-style).

Reply:
Having said that I tried hard thinking about better alternatives, but couldn't really come up with anything. Since it is typically type aliased and mostly intended for our crates I think it is OK not to think too much about the name here?